### PR TITLE
Urr

### PR DESCRIPTION
**Pull request by Roy-Hu**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,3 +19,5 @@ require (
 	github.com/wmnsk/go-pfcp v0.0.13
 	gopkg.in/yaml.v2 v2.4.0
 )
+
+replace github.com/free5gc/go-gtp5gnl => /home/free5gc/free5gc/go-gtp5gnl
```
Comment (khirono) [inline go.mod:23]: You had better remove it.

**Review by free5gc-org — COMMENTED** [review]
update go-gtp5gnl to v1.3.6

File: internal/pfcp/pfcp.go [hunk 2]
```diff
@@ -58,6 +58,7 @@ type PfcpServer struct {
 	rxTrans      map[string]*RxTransaction // key: RemoteAddr-Sequence
 	txSeq        uint32
 	log          *logrus.Entry
+	urrURSEQN    map[uint32]uint32
```
Comment (tim-ywliu) [inline internal/pfcp/pfcp.go:61]: what is `urrURSEQN` used for? It is not used in this PR. If no more use, please remove it.

File: internal/pfcp/pfcp.go [hunk 3]
```diff
@@ -75,6 +76,7 @@ func NewPfcpServer(cfg *factory.Config, driver forwarder.Driver) *PfcpServer {
 		txTrans:      make(map[string]*TxTransaction),
 		rxTrans:      make(map[string]*RxTransaction),
 		log:          logger.PfcpLog.WithField(logger.FieldListenAddr, listen),
+		urrURSEQN:    make(map[uint32]uint32),
```
Comment (tim-ywliu) [inline internal/pfcp/pfcp.go:79]: what is `urrURSEQN` used for? It is not used in this PR. If no more use, please remove it.

File: internal/pfcp/node.go [hunk 4]
```diff
@@ -218,11 +252,6 @@ func (s *Sess) UpdateBAR(req *ie.IE) error {
 }
 
 func (s *Sess) RemoveBAR(req *ie.IE) error {
-	err := s.rnode.driver.RemoveBAR(s.LocalID, req)
```
Comment (ianchen0119) [inline internal/pfcp/node.go:221]: Why remove this?

File: internal/forwarder/buff/server.go [hunk 5]
```diff
@@ -63,41 +69,109 @@ func (s *Server) Serve(wg *sync.WaitGroup) {
...
+		report_num := int(*(*uint16)(unsafe.Pointer(&b[off])))
+		off += 2
+		usars := []report.USAReport{}
+
+		for i := 0; i < report_num; i++ {
+			usar := report.USAReport{}
+
+			usar.URRID = (*(*uint32)(unsafe.Pointer(&b[off])))
+			off += 4
+			r := (*(*uint64)(unsafe.Pointer(&b[off])))
```
Comment (tim-ywliu) [inline internal/forwarder/buff/server.go:134]:
```
usar.USARTrigger.Flags = (*(*uint32)(unsafe.Pointer(&b[off])))
off += 4
```

File: internal/forwarder/buff/server.go [hunk 6]
```diff
@@ -63,41 +69,109 @@ func (s *Server) Serve(wg *sync.WaitGroup) {
...
 	seid := *(*uint64)(unsafe.Pointer(&b[off]))
 	off += 8
-	pdrid := *(*uint16)(unsafe.Pointer(&b[off]))
-	off += 2
-	action := *(*uint16)(unsafe.Pointer(&b[off]))
-	off += 2
-	return seid, pdrid, action, b[off:], nil
+
+	if msgtype == TYPE_URR_REPORT {
+		report_num := int(*(*uint16)(unsafe.Pointer(&b[off])))
```
Comment (tim-ywliu) [inline internal/forwarder/buff/server.go:125]:
```
		report_num := int(*(*uint32)(unsafe.Pointer(&b[off])))
		off += 4
```

File: internal/forwarder/buff/server.go (hunk 5 above)
Comment (tim-ywliu) [inline internal/forwarder/buff/server.go:134]:
```
			usar.USARTrigger.Flags = (*(*uint32)(unsafe.Pointer(&b[off])))
			off += 4
```

**Review by tim-ywliu — APPROVED** [review]

**Pull request merged**